### ISSUE TITLE
Mixed Outputs in Concurrent Requests via vLLM

### ISSUE THREAD

**Issue by Young-Chin** [Tencent-Hunyuan/HunyuanOCR#60]
I am experiencing an issue where outputs from concurrent requests are being mixed when using the HunyuanOCR model through vLLM.

Version & Command:
```bash
# vllm ==0.11.2.dev461+g8bbcf8b6e
vllm serve ./HunyuanOCR \
  --no-enable-prefix-caching \
  --mm-processor-cache-gb 0 \
  --gpu_memory_utilization 0.25
```
When sending concurrent requests via AsyncOpenAI, the generated text for one request occasionally contains fragments from other active requests within the same batch. This suggests potential interference between individual requests.
Example:
idx 391: "妈搭上这半条命" ← Correct for Image A
idx 392: "妈她求过这半条命什么" ← Corrupted
idx 393: "把她求回来做什么" Correct for Image C

I am unsure if this is a known issue within vLLM’s handling of concurrent requests, or if it stems from HunyuanOCR not yet fully supporting concurrent inference.

Any guidance or assistance would be greatly appreciated. Thank you.

**Comment by Kihensarn** [comment]
+1

**Comment by anker-c2** [comment]
I encountered the same while looking for a different problem in vLLM. Sounds like it's the same course, maybe try the nightly built. 
It's probably related to the same issue and related PR:
https://github.com/vllm-project/vllm/issues/29812

If it's not, it sounds like a vLLM issue to me regardless.

**Comment by Young-Chin** [comment]
> I encountered the same while looking for a different problem in vLLM. Sounds like it's the same course, maybe try the nightly built. It's probably related to the same issue and related PR: [vllm-project/vllm#29812](https://github.com/vllm-project/vllm/issues/29812)
> 
> If it's not, it sounds like a vLLM issue to me regardless.

Well, I think this is the Incompatibility between HunyuanOCR and vLLM. They have also identified this issue and will follow up to address it.

**Comment by vince62s** [comment]
if I am not mistaken, the reason is that when there is several images in a batch, they reshape each image to have (N, D) N being the number of patches, D the hidden dim (1152) then they concat across that dimension when sending it to Vit. Indeed this is incorrect.

in HF's code it's here:
```
class HunYuanVisionTransformer(nn.Module):
    config: HunYuanVLVisionConfig
    _no_split_modules = ["HunYuanVLVisionBlock"]

    def __init__(self, config: HunYuanVLVisionConfig):
        super().__init__()
        self.config = config
        self.embeddings = HunYuanVisionPatchEmbed(config)
        self.layers = nn.ModuleList([HunYuanVisionBlock(config) for _ in range(config.num_hidden_layers)])
        self.perceive = HunYuanVisionPatchMerger(
            self.config.hidden_size,
            self.config.text_hidden_size,
            self.config.spatial_merge_size,
            self.config.rms_norm_eps,
        )

    def get_activation_function(self, act_name: str):
        act_map = {
            "gelu": nn.GELU(),
            "relu": nn.ReLU(),
            "silu": nn.SiLU(),
        }
        return act_map.get(act_name.lower(), nn.GELU())  # default GELU

    # @auto_docstring
    def forward(
        self,
        x: torch.Tensor,
        grid_thw: list[list[int]],
    ) -> torch.Tensor:
        #
        r"""
        grid_thw (`torch.LongTensor` of shape `(num_images, 3)`):
            The temporal, height and width dimensions of feature shape for each image. Each row contains [t, h, w] values.
        """
        hidden_states = self.embeddings(x, grid_thw)
        for layer in self.layers:
            hidden_states = layer(hidden_states)
```

x is shaped (1, sum of N, D)

I think the only way is either to process image by image, to pad small images to batch them all to the same number of patches, OR better to use a diagonal mask to segregate the images (same as pixtral / mistral)

**Comment by anker-c2** [comment]
Actually i was now able to reproduce the issue, once again it only happens under load (which is to be expected giving the nature of the bug being in the batching). It was not fixed with the vLLM PR as i was hoping. 

@vince62s Did you actually manage to make some progress on the Issue, or should i open a Issue on vLLM?

**Comment by vince62s** [comment]
it works in my code now: https://github.com/eole-nlp/eole/blob/main/eole/encoders/vision.py#L242-L245
but I'm not using vLLM. It should be straight forward.

**Comment by vince62s** [comment]
in fact in vLLM it requires a little bit more of work because they did not implement the visionblock fully, it does not accept a mask in the forward.
here:
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/hunyuan_vision.py#L519
there should be a mask created like here:
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/pixtral.py#L855-L857


But this code here does not accept a mask:
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/hunyuan_vision.py#L291-L296
nor here:
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/hunyuan_vision.py#L243-L251

**Comment by Young-Chin** [comment]
> in fact in vLLM it requires a little bit more of work because they did not implement the visionblock fully, it does not accept a mask in the forward. here: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/hunyuan_vision.py#L519 there should be a mask created like here: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/pixtral.py#L855-L857
> 
> But this code here does not accept a mask: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/hunyuan_vision.py#L291-L296 nor here: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/hunyuan_vision.py#L243-L251

Following your approach, I added the attention mask to the layer: https://github.com/vllm-project/vllm/blob/f6227c22ab8976a24913122874c24624102da1b4/vllm/model_executor/models/hunyuan_vision.py#L519 
and pass the argument from transformer to attention. 

Since MultiHeadAttention in vLLM does not support attention mask, I rewrote the forward part to replace self.attn(). 
Eventually no more mixed results here, but the OCR accuracy seems to have decreased compared to image-by-image processing. 

Maybe there is something wrong with the rewritten forward：
 ```
   def forward(
        self,
        x: torch.Tensor,
        attention_mask,
    ) -> torch.Tensor:

        qkv, _ = self.qkv(x)
        q, k, v = qkv.chunk(3, dim=-1)
        num_heads = self.num_attention_heads_per_partition
        head_dim = self.hidden_size_per_attention_head
        
        q = q.view(batch_size, seq_len, num_heads, head_dim).transpose(1, 2)
        k = k.view(batch_size, seq_len, num_heads, head_dim).transpose(1, 2)
        v = v.view(batch_size, seq_len, num_heads, head_dim).transpose(1, 2)
        out = F.scaled_dot_product_attention(q, k, v, attn_mask=attention_mask)
        out = out.transpose(1, 2).contiguous().reshape(batch_size, seq_len, -1)
        output, _ = self.o_proj(out)
        return output
```

**Comment by vince62s** [comment]
did you try image by image with your modified code ? is it also different?

**Comment by Young-Chin** [comment]
> did you try image by image with your modified code ? is it also different?

Yep, processing images one by one is also different. I guess my rewritten part is different from the original implementation.

**Comment by vince62s** [comment]
As I said on the vLLM repo, this is not the only issue, the HF/vLLM implementation of the position_ids preparation is broken (not batch compatible) which means the xdrope does not work properly on top of this one too (#49). In essence it does not prevent from working "OK" but this is not the intended way it should work. Even with the accurate way to compute xdrope (I pushed it last night in my repo) does not make things 100% perfect. I just hope the model was trained with accurate code for xdrope.

**Comment by anker-c2** [comment]
I built a small hotpatch for the xdrope issue, however i didn't put much effort into measuring the precision. 
Since i am just using the AsyncOpenAI client with the vLLM http server frontend for all my intends and purposes, it doesn't matter. 

It "only" becomes a problem, when using vLLM directly (and therefore directly making batched calls with multiple prompts / images). When using the http frontend (e.g. with docker) with the OAI client, each requst made goes through the processor independently with batch_size 1. Therefore the issue doesn't affect images being batched together by the vllm scheduler after the fact. 

@vince62s thanks a lot for all of your work and insights on the issues, I didn't even notice the xdrope issue. Highly appreciated your help and comment on the vllm PR!

Edit: So as a workaround the nightly vllm-openai build via HTTP frontend (with the openai client or similiar) should now work as expected. At least for me it works exactly as it should, the quality is as expected.

**Comment by iscpatrol-rgb** [comment]
<img width="1467" height="601" alt="Image" src="https://github.com/user-attachments/assets/bb86d4c1-3f09-48a0-bf56-caf3f006778b" />
我不知道如何重现这个问题，有一次在https://huggingface.co/spaces/tencent/HunyuanOCR做测试的时候，发现返回的结果混合了上次识别的结果
